src/mesh.cpp

```diff
@@ -783,6 +829,517 @@ RegularMesh::count_sites(const std::vector<Particle::Bank>& bank,
+
+  // Calculate number of surface crossings
+  int n_cross = 0;
+  for (int i = 0; i < 3; ++i) n_cross += std::abs(ijk1[i] - ijk0[i]);
+  if (n_cross == 0) return;
+
+  // Bounding coordinates
+  Position xyz_cross;
+  for (int i = 0; i < 3; ++i) {
+    if (u[i] > 0.0) {
+      xyz_cross[i] = grid_[i][ijk0[i]];
+    } else {
+      xyz_cross[i] = grid_[i][ijk0[i] - 1];
+    }
+  }
+
+  for (int j = 0; j < n_cross; ++j) {
+    // Set the distances to infinity
+    Position d {INFTY, INFTY, INFTY};
+
+    // Determine closest bounding surface. We need to treat
+    // special case where the cosine of the angle is zero since this would
+    // result in a divide-by-zero.
+    double distance = INFTY;
+    for (int i = 0; i < 3; ++i) {
+      if (u[i] == 0) {
+        d[i] = INFTY;
+      } else {
+        d[i] = (xyz_cross[i] - r0[i])/u[i];
+      }
+      distance = std::min(distance, d[i]);
+    }
+
+    // Loop over the dimensions
+    for (int i = 0; i < 3; ++i) {
+      // Check whether distance is the shortest distance
+      if (distance == d[i]) {
+
+        // Check whether particle is moving in positive i direction
+        if (u[i] > 0) {
+
+          // Outward current on i max surface
+          int i_surf = 4*i + 3;
+          int i_mesh = get_bin_from_indices(ijk0);
+          int i_bin = 4*3*i_mesh + i_surf - 1;
+          bins.push_back(i_bin);
+
+          // Advance position
+          ++ijk0[i];
+          xyz_cross[i] = grid_[i][ijk0[i]];
+
+          // Inward current on i min surface
+          i_surf = 4*i + 2;
+          i_mesh = get_bin_from_indices(ijk0);
+          i_bin = 4*3*i_mesh + i_surf - 1;
+          bins.push_back(i_bin);
+
+        } else {
+          // The particle is moving in the negative i direction
+
+          // Outward current on i min surface
+          int i_surf = 4*i + 1;
+          int i_mesh = get_bin_from_indices(ijk0);
+          int i_bin = 4*3*i_mesh + i_surf - 1;
+          bins.push_back(i_bin);
+
+          // Advance position
+          --ijk0[i];
+          xyz_cross[i] = grid_[i][ijk0[i] - 1];
+
+          // Inward current on i min surface
+          i_surf = 4*i + 4;
+          i_mesh = get_bin_from_indices(ijk0);
+          i_bin = 4*3*i_mesh + i_surf - 1;
+          bins.push_back(i_bin);
+        }
+      }
+    }
+
+    // Calculate new coordinates
+    r0 += distance * u;
+  }
+}
+
+int RectilinearMesh::get_bin(Position r) const
+{
+  // Determine indices
+  int ijk[3];
+  bool in_mesh;
+  get_indices(r, ijk, &in_mesh);
+  if (!in_mesh) return -1;
+
+  // Convert indices to bin
+  return get_bin_from_indices(ijk);
+}
+
+int RectilinearMesh::get_bin_from_indices(const int* ijk) const
+{
+  return ((ijk[2] - 1)*shape_[1] + (ijk[1] - 1))*shape_[0] + ijk[0] - 1;
+}
+
+void RectilinearMesh::get_indices(Position r, int* ijk, bool* in_mesh) const
+{
+  *in_mesh = true;
+
+  for (int i = 0; i < 3; ++i) {
+    if (r[i] < grid_[i].front() || r[i] > grid_[i].back()) {
```

Comment:
I noticed that `RectilinearMesh` assumes (reasonably) that divisions along each axes are increasing. I tried throwing in some decreasing values for the plot test and it worked correctly without complaint, but I think we may run into issues here (and maybe other places as well). It'd probably be best to either make this clear in the docs or provide a check when creating the mesh.

Reply:
Sure thing.  I added a check in the `RectilinearMesh` constructor